### ISSUE TITLE
PerlIO::scalar doesn't respect readonly-ness

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8583]
Migrated from [rt.perl.org#40267](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40267) (status was 'resolved')

Searchable as RT40267$

**Comment by p5pRT** [comment]
### From blgl@hagernas.com

#### Created by blgl@hagernas.com


The example program below generates no errors or warnings
and outputs "extra everything"\.  \(Yes\, this bug is a whopper\.\)

\#\! /usr/local/bin/perl
use strict;
use warnings;
open\(my $fh\,">"\,\\undef\)
&nbsp;     or die "open&#8203;: $\!";
print \{$fh\} "extra everything\\n"
&nbsp;     or die "print&#8203;: $\!";
close\($fh\)
&nbsp;     or die "close&#8203;: $\!";
print undef;

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl v5.8.8:

Configured by blgl at Fri Jun  9 19:26:53 CEST 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
   Platform:
     osname=darwin, osvers=8.6.0, archname=darwin-thread- 
multi-64int-2level
     uname='darwin sunwukung.sth.ownit.se 8.6.0 darwin kernel version  
8.6.0: tue mar 7 16:58:48 pst 2006;  
root:xnu-792.6.70.obj~1release_ppc power macintosh powerpc '
     config_args=''
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define  
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=define use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-common -DPERL_DARWIN -no-cpp-precomp - 
fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/ 
include',
     optimize='-Os',
     cppflags='-no-cpp-precomp -fno-common -DPERL_DARWIN -no-cpp- 
precomp -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/ 
usr/local/include'
     ccversion='', gccversion='4.0.1 (Apple Computer, Inc. build  
5247)', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=87654321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long long', ivsize=8, nvtype='double', nvsize=8,  
Off_t='off_t', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -L/usr/ 
local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-lc
     perllibs=-lc
     libc=/usr/lib/libc.dylib, so=dylib, useshrplib=false,  
libperl=libperl.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/ 
usr/local/lib'

Locally applied patches:



@INC for perl v5.8.8:
     /usr/local/lib/perl5/5.8.8/darwin-thread-multi-64int-2level
     /usr/local/lib/perl5/5.8.8
     /usr/local/lib/perl5/site_perl/5.8.8/darwin-thread- 
multi-64int-2level
     /usr/local/lib/perl5/site_perl/5.8.8
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.8.8:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/blgl
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/Users/blgl/bin:/Users/blgl/bin/powerpc-apple-darwin:/usr/ 
local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin
     PERL_BADLANG (unset)
     SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From mjtg@cam.ac.uk

Bo Lindbergh \(via RT\) \<perlbug\-followup@&#8203;perl\.org> wrote
> The example program below generates no errors or warnings
> and outputs "extra everything"\.  \(Yes\, this bug is a whopper\.\)

And anyone fixing this should deal with these variations&#8203;:

perl \-wl
open\(my $fh\,">"\,\\\(\!0\)\)
&nbsp;     or die "open&#8203;: $\!";
print \{$fh\} "extra everything\\n"
&nbsp;     or die "print&#8203;: $\!";
close\($fh\)
&nbsp;     or die "close&#8203;: $\!";
print \!0;
print \!\!0
\_\_END\_\_
extra everything

perl \-wl
my $x = \\3\.14;
open\(my $fh\,">"\,$x\)
&nbsp;     or die "open&#8203;: $\!";
print \{$fh\} "extra everything\\n"
&nbsp;     or die "print&#8203;: $\!";
close\($fh\)
&nbsp;     or die "close&#8203;: $\!";
print $$x;
\_\_END\_\_
extra everything


Mike Guy

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgarcia

On 02/09/06\, via RT Bo Lindbergh \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> The example program below generates no errors or warnings
> and outputs "extra everything"\.  \(Yes\, this bug is a whopper\.\)
>
> \#\! /usr/local/bin/perl
> use strict;
> use warnings;
> open\(my $fh\,">"\,\\undef\)
>      or die "open&#8203;: $\!";
> print \{$fh\} "extra everything\\n"
>      or die "print&#8203;: $\!";
> close\($fh\)
>      or die "close&#8203;: $\!";
> print undef;

Thanks for the report; I've fixed that in bleadperl by change \#28798\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
